File: wrappers/python/tests/demo/anoncreds.py
from indy import anoncreds, wallet

from tests.utils import storage

import pytest
import logging
import json

logging.basicConfig(level=logging.DEBUG)


@pytest.mark.asyncio
async def test_anoncreds_demo_works(cleanup_storage):
    pool_name = "anoncreds_demo_pool"
    wallet_name = "anoncreds_demo_wallet"

    # 1. Create My Wallet and Get Wallet Handle
    await wallet.create_wallet(pool_name, wallet_name, None, None, None)
    wallet_handle = await wallet.open_wallet(wallet_name, None, None)

    # 2. Issuer create Claim Definition for Schema
    schema_seq_no = 1
    schema = {
        'seqNo': schema_seq_no,
        'data': {
            'name': 'gvt',
            'version': '1.0',
            'keys': ['age', 'sex', 'height', 'name']
        }
    }
    schema_json = json.dumps(schema)
    issuer_did = 'NcYxiDXkpYi6ov5FcYDi1e'

    claim_def_json = \
        await anoncreds.issuer_create_and_store_claim_def(wallet_handle, issuer_did, schema_json, 'CL', False)

    # 3. Prover create Master Secret
    master_secret_name = 'master_secret'
    await anoncreds.prover_create_master_secret(wallet_handle, master_secret_name)

    # 4. Prover create Claim Request
    prover_did = 'BzfFCYk'
    claim_offer = {
        'issuer_did': issuer_did,
        'schema_seq_no': schema_seq_no
    }

    claim_req_json = await anoncreds.prover_create_and_store_claim_req(wallet_handle, prover_did,
                                                                       json.dumps(claim_offer),
                                                                       claim_def_json, master_secret_name)

    # 5. Issuer create Claim for Claim Request
    claim = {
        'sex': ['male', '5944657099558967239210949258394887428692050081607692519917050011144233115103'],
        'name': ['Alex', '1139481716457488690172217916278103335'],
        'height': ['175', '175'],
        'age': ['28', '28']
    }

    (_, claim_json) = await anoncreds.issuer_create_claim(wallet_handle, claim_req_json, json.dumps(claim), -1, -1)

    # 6. Prover process and store Claim
    await anoncreds.prover_store_claim(wallet_handle, claim_json)

    # 7. Prover gets Claims for Proof Request
    proof_req = {
        'nonce': '123432421212',
        'name': 'proof_req_1',
        'version': '0.1',
        'requested_attrs': {
            'attr1_uuid': {'schema_seq_no': schema_seq_no, 'name': 'name'}
        },
        'requested_predicates': {
            'predicate1_uuid': {'attr_name': 'age', 'p_type': 'GE', 'value': 18}
        }
    }
    proof_req_json = json.dumps(proof_req)

    claim_for_proof_json = await anoncreds.prover_get_claims_for_proof_req(wallet_handle, proof_req_json)
    claims_for_proof = json.loads(claim_for_proof_json)

    claim_for_attr1 = claims_for_proof['attrs']['attr1_uuid']
    assert 1 == len(claim_for_attr1)

    claim_uuid = claim_for_attr1[0]['claim_uuid']

    # 8. Prover create Proof for Proof Request
    requested_claims = {
        'self_attested_attributes': {},
        'requested_attrs': {'attr1_uuid': [claim_uuid, True]},
        'requested_predicates': {'predicate1_uuid': claim_uuid}
    }
    requested_claims_json = json.dumps(requested_claims)

    schemas_json = json.dumps({claim_uuid: schema})
    claim_defs_json = json.dumps({claim_uuid: json.loads(claim_def_json)})
    revoc_regs_json = json.dumps({})

    proof_json = await anoncreds.prover_create_proof(wallet_handle, proof_req_json, requested_claims_json, schemas_json,
                                                     master_secret_name, claim_defs_json, revoc_regs_json)
    proof = json.loads(proof_json)

    assert 'Alex' == proof['requested_proof']['revealed_attrs']['attr1_uuid'][1]

    # 9. Verifier verify proof
    assert await anoncreds.verifier_verify_proof(proof_req_json, proof_json, schemas_json, claim_defs_json,
                                                 revoc_regs_json)

    # 10. Close wallet
    await wallet.close_wallet(wallet_handle)

File: samples/java/src/main/java/org/hyperledger/indy/sample/Tests/Anoncreds.java
package org.hyperledger.indy.sample.Tests;

import org.hyperledger.indy.sdk.anoncreds.AnoncredsResults;
import org.hyperledger.indy.sdk.pool.Pool;
import org.hyperledger.indy.sdk.pool.PoolJSONParameters;
import org.hyperledger.indy.sdk.wallet.Wallet;
import org.json.JSONArray;
import org.json.JSONObject;
import org.hyperledger.indy.sample.utils.PoolUtils;
import org.hyperledger.indy.sample.utils.StorageUtils;

import static org.hyperledger.indy.sdk.anoncreds.Anoncreds.*;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;


public class Anoncreds {

	public static void run() throws Exception {

		StorageUtils.cleanupStorage();

		//1. Create and Open Pool
		String poolName = PoolUtils.createPoolLedgerConfig();

		PoolJSONParameters.OpenPoolLedgerJSONParameter config2 = new PoolJSONParameters.OpenPoolLedgerJSONParameter(null, null, null);
		Pool pool = Pool.openPoolLedger(poolName, config2.toJson()).get();

		//2. Issuer Create and Open Wallet
		Wallet.createWallet(poolName, "issuerWallet", "default", null, null).get();
		Wallet issuerWallet = Wallet.openWallet("issuerWallet", null, null).get();

		//3. Prover Create and Open Wallet
		Wallet.createWallet(poolName, "proverWallet", "default", null, null).get();
		Wallet proverWallet = Wallet.openWallet("proverWallet", null, null).get();

		//4. Issuer create ClaimDef
		String schemaJson = "{\n" +
				"                    \"seqNo\":1,\n" +
				"                    \"data\": {\n" +
				"                        \"name\":\"gvt\",\n" +
				"                        \"version\":\"1.0\",\n" +
				"                        \"keys\":[\"age\",\"sex\",\"height\",\"name\"]\n" +
				"                    }\n" +
				"                }";
		String issuerDid = "NcYxiDXkpYi6ov5FcYDi1e";

		String claimDef = issuerCreateAndStoreClaimDef(issuerWallet, issuerDid, schemaJson, null, false).get();
		assertNotNull(claimDef);

		//5. Prover create Master Secret
		String masterSecret = "masterSecretName";
		proverCreateMasterSecret(proverWallet, masterSecret).get();

		//6. Prover store Claim Offer
		String claimOffer = String.format("{\"issuer_did\":\"%s\", \"schema_seq_no\":%d}", issuerDid, 1);
		proverStoreClaimOffer(proverWallet, claimOffer).get();

		//7. Prover get Claim Offers
		String claimOfferFilter = String.format("{\"issuer_did\":\"%s\"}", issuerDid);
		String claimOffersJson = proverGetClaimOffers(proverWallet, claimOfferFilter).get();

		JSONArray claimOffersObject = new JSONArray(claimOffersJson);
		assertEquals(claimOffersObject.length(), 1);

		JSONObject claimOfferObject = claimOffersObject.getJSONObject(0);
		String claimOfferJson = claimOfferObject.toString();

		//8. Prover create ClaimReq
		String proverDid = "BzfFCYk";
		String claimReq = proverCreateAndStoreClaimReq(proverWallet, proverDid, claimOfferJson, claimDef, masterSecret).get();
		assertNotNull(claimReq);

		//9. Issuer create Claim
		String claimAttributesJson = "{\n" +
				"               \"sex\":[\"male\",\"5944657099558967239210949258394887428692050081607692519917050011144233115103\"],\n" +
				"               \"name\":[\"Alex\",\"1139481716457488690172217916278103335\"],\n" +
				"               \"height\":[\"175\",\"175\"],\n" +
				"               \"age\":[\"28\",\"28\"]\n" +
				"        }";

		AnoncredsResults.IssuerCreateClaimResult createClaimResult = issuerCreateClaim(issuerWallet, claimReq, claimAttributesJson, - 1).get();
		assertNotNull(createClaimResult);
		String claimJson = createClaimResult.getClaimJson();

		//10. Prover store Claim
		proverStoreClaim(proverWallet, claimJson).get();

		//11. Prover gets Claims for Proof Request
		String proofRequestJson = "{\n" +
				"                          \"nonce\":\"123432421212\",\n" +
				"                          \"name\":\"proof_req_1\",\n" +
				"                          \"version\":\"0.1\",\n" +
				"                          \"requested_attrs\":{\"attr1_uuid\":{\"schema_seq_no\":1,\"name\":\"name\"},\n" +
				"                                                \"attr2_uuid\":{\"schema_seq_no\":1,\"name\":\"sex\"}},\n" +
				"                          \"requested_predicates\":{\"predicate1_uuid\":{\"attr_name\":\"age\",\"p_type\":\"GE\",\"value\":18}}\n" +
				"                  }";

		String claimsForProofJson = proverGetClaimsForProofReq(proverWallet, proofRequestJson).get();
		assertNotNull(claimsForProofJson);

		JSONObject claimsForProof = new JSONObject(claimsForProofJson);
		JSONArray claimsForAttribute1 = claimsForProof.getJSONObject("attrs").getJSONArray("attr1_uuid");
		JSONArray claimsForAttribute2 = claimsForProof.getJSONObject("attrs").getJSONArray("attr1_uuid");
		JSONArray claimsForPredicate = claimsForProof.getJSONObject("predicates").getJSONArray("predicate1_uuid");

		assertEquals(claimsForAttribute1.length(), 1);
		assertEquals(claimsForAttribute2.length(), 1);
		assertEquals(claimsForPredicate.length(), 1);

		String claimUuid = claimsForAttribute1.getJSONObject(0).getString("claim_uuid");

		//12. Prover create Proof
		String selfAttestedValue = "yes";
		String requestedClaimsJson = String.format("{\n" +
				"                                          \"self_attested_attributes\":{\"self1\":\"%s\"},\n" +
				"                                          \"requested_attrs\":{\"attr1_uuid\":[\"%s\", true],\n" +
				"                                                               \"attr2_uuid\":[\"%s\", false]},\n" +
				"                                          \"requested_predicates\":{\"predicate1_uuid\":\"%s\"}\n" +
				"                                        }", selfAttestedValue, claimUuid, claimUuid, claimUuid);

		String schemasJson = String.format("{\"%s\":%s}", claimUuid, schemaJson);
		String claimDefsJson = String.format("{\"%s\":%s}", claimUuid, claimDef);
		String revocRegsJson = "{}";


		String proofJson = proverCreateProof(proverWallet, proofRequestJson, requestedClaimsJson, schemasJson,
				masterSecret, claimDefsJson, revocRegsJson).get();
		assertNotNull(proofJson);

		JSONObject proof = new JSONObject(proofJson);

		//13. Verifier verify Proof
		assertEquals("Alex",
				proof.getJSONObject("requested_proof").getJSONObject("revealed_attrs").getJSONArray("attr1_uuid").getString(1));

		assertNotNull(proof.getJSONObject("requested_proof").getJSONObject("unrevealed_attrs").getString("attr2_uuid"));

		assertEquals(selfAttestedValue, proof.getJSONObject("requested_proof").getJSONObject("self_attested_attrs").getString("self1"));

		Boolean valid = verifierVerifyProof(proofRequestJson, proofJson, schemasJson, claimDefsJson, revocRegsJson).get();
		assertTrue(valid);

		issuerWallet.closeWallet().get();
		Wallet.deleteWallet("issuerWallet", null).get();

		proverWallet.closeWallet().get();
		Wallet.deleteWallet("proverWallet", null).get();

		pool.closePoolLedger().get();

		StorageUtils.cleanupStorage();
	}
}
